unit-tests-java8/src/test/java/org/eclipse/collections/test/map/mutable/MutableMapIterableTestCase.java

```diff
@@ -0,0 +1,234 @@
+                    map2);
+        }
+    }
+
+    @Test
+    default void Map_putAll()
+    {
+        MutableMapIterable<Integer, String> map = this.newWithKeysValues(
+                3, "Three",
+                2, "2");
+        MutableMapIterable<Integer, String> toAdd = this.newWithKeysValues(
+                2, "Two",
+                1, "One");
+
+        map.putAll(toAdd);
+
+        MutableMapIterable<Integer, String> expected = this.newWithKeysValues(
+                3, "Three",
+                2, "Two",
+                1, "One");
+
+        assertEquals(expected, map);
+
+        //Testing JDK map
+        MutableMapIterable<Integer, String> map2 = this.newWithKeysValues(
+                3, "Three",
+                2, "2");
+        HashMap<Integer, String> hashMaptoAdd = new LinkedHashMap<>();
+        hashMaptoAdd.put(2, "Two");
+        hashMaptoAdd.put(1, "One");
+        map2.putAll(hashMaptoAdd);
+
+        assertEquals(expected, map2);
+    }
+
+    @Test
+    default void MutableMapIterable_getIfAbsentPut()
+    {
+        MutableMapIterable<String, Integer> map = this.newWithKeysValues("3", 3, "2", 2, "1", 1);
+        assertEquals(3, map.getIfAbsentPut("3", () -> {
+            throw new AssertionError();
+        }));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1), map);
+
+        assertEquals(4, map.getIfAbsentPut("4", () -> 4));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1, "4", 4), map);
+
+        MutableMapIterable<String, Integer> map2 = this.newWithKeysValues("3", 3, "2", 2, "1", 1);
+        assertEquals(3, map2.getIfAbsentPut("3", 4));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1), map2);
+
+        assertEquals(4, map2.getIfAbsentPut("4", 4));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1, "4", 4), map2);
+
+        MutableMapIterable<String, Integer> map3 = this.newWithKeysValues("3", 3, "2", 2, "1", 1);
+        assertEquals(3, map3.getIfAbsentPutWithKey("3", key -> {
+            throw new AssertionError();
+        }));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1), map3);
+
+        assertEquals(14, map3.getIfAbsentPutWithKey("4", key -> Integer.parseInt(key) + 10));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1, "4", 14), map3);
+
+        // TODO: This method is strange. Should we deprecate it? I could see a hyptothetical getIfAbsentPutWithKeyWith being used, but I'm having trouble imagining a use-case for getIfAbsentPutWith.
+
+        MutableMapIterable<String, Integer> map4 = this.newWithKeysValues("3", 3, "2", 2, "1", 1);
+        assertEquals(3, map4.getIfAbsentPutWith("3", x -> x + 10, 4));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1), map4);
+
+        assertEquals(14, map4.getIfAbsentPutWith("4", x -> x + 10, 4));
+        assertEquals(this.newWithKeysValues("3", 3, "2", 2, "1", 1, "4", 14), map4);
```

Comment:
Usually the *With methods are alternatives that take a Function2 instead of a Function. Here, the With method takes a Function instead of a Function0. I'm struggling to think of an example when it would be useful. If there are examples, that's fine, otherwise I think we should consider deprecating the method.

Reply:
There are real use cases where the key can be used to create the value. This happens when the value needs to keep a reference to its key (the code looks like ```x -> new Foo(x)```). 1 see one usage in real code.

Conceptually, there is another usage, where the key can be used to lookup an initial value for the value. If the rest of the context is static, the lambda can also remain static.

Reply:
To clarify, the comment is about getIfAbsentPutWith, not getIfAbsentPutWithKey. It doesn't get access to the key in its lambda. It only gets access to the param, the second argument. A hypothetical new method getIfAbsentPutWithKeyWith (taking both the key and a param) makes more sense to me. I never really paid attention to this method until I was trying to cover it in a test.

Reply:
```getIfAbsentPutWith``` is quite useful on its own. The typical example is aggregation where the sum object is different from the individual value objects. For plain numbers, that would be MutableInteger vs Integer, etc. For more complex domains, the sum object may have more information, but they almost always need the current value.

Reply:
Not sure if I follow, so I'll add some code examples.

I think the form of `getIfAbsentPut` that takes a `Function` is useful for aggregations.
```java
Counter counter = this.items.getIfAbsentPut(item, Counter::new);
counter.increment();
```

I don't understand how `getIfAbsentPutWith` can be useful because its function takes *neither* the key nor the value.

```
<P> V getIfAbsentPutWith(K key, Function<? super P, ? extends V> function, P parameter);
```

All that said, I looked through every usage of getIfAbsentPutWith in the library, and I found one usage I wrote myself in AbstractMutableMultimap.

```
    private C getIfAbsentPutCollection(K key)
    {
        return this.map.getIfAbsentPutWith(key, this.createCollectionBlock(), this);
    }
```

I'll remove this comment from the code. Thanks for the feedback.